### ISSUE TITLE
Bug: exception_handlers are not applied when debug flag is False

### ISSUE THREAD

**Issue by TheZwieback** [litestar-org/litestar#3579]
### Description

I want to configure a custom exception handler for the following use case. I want to run the app in production but log stack traces of internal server errors but other HTTPException should not be logged with stack trace as they are usually triggered specifically and converted to a proper response by Litestar.
In this way I can raise HTTPExceptions for issues inside my code without having to create explicit responses. Also I have the case that I want to trigger HTTPExceptions in a dependency injection directly in one place without having to duplicate error handling in each endpoint where it is used.
Anyways, the current issue is that I cannot get the exception handler to trigger when debug is set to False in the app.

### MCVE

```python
from litestar.exceptions import HTTPException
from litestar.status_codes import HTTP_500_INTERNAL_SERVER_ERROR
from litestar import Request, Response

from litestar import Litestar, post
import logging
from litestar.exceptions import LitestarException


def exception_handler(_: Request, exc: Exception) -> Response:
    if isinstance(exc, HTTPException):
        status_code = getattr(exc, "status_code", HTTP_500_INTERNAL_SERVER_ERROR)
        detail = getattr(exc, "detail", "")

        return Response(
            content={"detail": detail},
            status_code=status_code,
        )

    logging.exception(exc)
    status_code = HTTP_500_INTERNAL_SERVER_ERROR
    detail = {"special_detail": "Internal Server Error"}

    return Response(content=detail, status_code=status_code)


@post(
    "/test",
)
async def test() -> Response[str]:
    1 / 0
    return Response("test")


app = Litestar(
    route_handlers=[test],
    exception_handlers={LitestarException: exception_handler},
)
```

### Steps to reproduce

```bash
1. Run the code
2. Trigger the endpoint
3. response body shows detail property instead of special_detail
```

### Screenshots

```bash
"![SCREENSHOT_DESCRIPTION](SCREENSHOT_LINK.png)"
```

### Logs

```bash
INFO:     Started server process [48233]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:38254 - "GET /schema/swagger HTTP/1.1" 200 OK
INFO:     127.0.0.1:38254 - "POST /test HTTP/1.1" 500 Internal Server Error
```

### Litestar Version

2.9.0

### Platform

- [X] Linux
- [ ] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

**Comment by Alc-Alc** [comment]
Not sure I understood this, you are applying the exception handler to `LitestarException` but your route handler raises a different exception, so how is it you expect it to trigger?

Your MCVE never applies the handler even if debug = True for the same reason stated above

**Comment by TheZwieback** [comment]
Thanks, I see my mistake. I assumed the LitestarException map would catch all exceptions for some reason. I checked the code and if use the correct error mapping it works. 
Let me rephrase my use case, I want to log stacktraces for any error that is not an HTTPException with 4XX status code. For this I want to implement an exception_handler that creates a regular 500 response without leaking the stacktrace as it does when debug is true and not printing stacktraces for any "known" error.
